internal/npm_package/packager.js

```diff
@@ -97,43 +97,49 @@ function main(args) {
-        `generated file should belong in 'deps' instead.`);
+    if (src.startsWith('external/')) {
+      // If srcs is from external workspace drop the external/wksp portion
+      copyWithReplace(
+          src, path.join(outDir, src.split('/').slice(2).join('/')), replacements,
+          renameBuildFiles);
+    } else {
+      // Source is from local workspace
+      if (baseDir && !src.startsWith(`${baseDir}/`)) {
+        throw new Error(
+            `${src} in 'srcs' does not reside in the base directory, ` +
+            `generated file should belong in 'deps' instead.`);
+      }
+      copyWithReplace(
+          src, path.join(outDir, path.relative(baseDir, src)), replacements, renameBuildFiles);
     }
-    copyWithReplace(
-        src, path.join(outDir, path.relative(baseDir, src)), replacements, renameBuildFiles);
   }
 
   function outPath(f) {
-    function findRoot() {
-      for (ext of vendorExternalArg.split(',').filter(s => !!s)) {
-        const candidate = path.join(binDir, 'external', ext);
-        if (!path.relative(candidate, f).startsWith('..')) {
-          return candidate;
-        }
-      }
-      if (!path.relative(binDir, f).startsWith('..')) {
-        return binDir;
-      } else if (!path.relative(genDir, f).startsWith('..')) {
-        return genDir;
-      } else {
-        // It might be nice to enforce here that deps don't contain sources
-        // since those belong in srcs above.
-        // The `deps` attribute should typically be outputs of other rules.
-        // However, things like .d.ts sources of a ts_library or data attributes
-        // of ts_library will result in source files that appear in the deps
-        // so we have to allow this.
-        return '.';
+    for (ext of vendorExternalArg.split(',').filter(s => !!s)) {
+      const candidate = path.join(binDir, 'external', ext);
+      if (!path.relative(candidate, f).startsWith('..')) {
+        return path.join(outDir, path.relative(candidate, f));
       }
     }
-    return path.join(outDir, path.relative(path.join(findRoot(), baseDir), f));
+    if (!path.relative(binDir, f).startsWith('..')) {
+      return path.join(outDir, path.relative(path.join(binDir, baseDir), f));
+    } else if (!path.relative(genDir, f).startsWith('..')) {
+      return path.join(outDir, path.relative(path.join(genDir, baseDir), f));
+    } else {
+      // It might be nice to enforce here that deps don't contain sources
+      // since those belong in srcs above.
+      // The `deps` attribute should typically be outputs of other rules.
+      // However, things like .d.ts sources of a ts_library or data attributes
+      // of ts_library will result in source files that appear in the deps
+      // so we have to allow this.
+      return path.join(outDir, path.relative(baseDir, f));
+    }
   }
 
   // deps like bazel-bin/baseDir/my/path is copied to outDir/my/path
   // Don't include external directories in the package, these should be installed
```

Comment:
is this comment out-of-date now?

Reply:
Yes. Good catch. external repo files may now be included if the external repos are explicitly listed in `vendor_external`. This is handled in `npm_package.bzl`:
```
# Takes a depset of files and returns a corresponding list of file paths without any files
# that aren't part of the specified package path. Also include files from external repositories
# that explicitly specified in the vendor_external list.
def _filter_out_external_files(ctx, files, package_path):
    result = []
    for file in files:
        if file.short_path.startswith(package_path):
            result.append(file.path)
        else:
            for v in ctx.attr.vendor_external:
                if file.short_path.startswith("../%s/" % v):
                    result.append(file.path)
    return result
```
I hijacked that attribute you put in originally to support vendoring in the rules_typescript code.

Reply:
Necessary to support the `npm_bazel_foobar` local_repository pattern where the rules themselves are. The npm_package for karma, for example, makes use of this:
```
npm_package(
    name = "npm_package",
    srcs = [
        "@npm_bazel_karma//:package_contents",
    ],
    replacements = COMMON_REPLACEMENTS,
    vendor_external = [
        "npm_bazel_karma",
    ],
    deps = [
        ":copy_LICENSE",
        ":generate_README",
        "@npm_bazel_karma//:bazel_karma_lib",
    ],
)
```